### DISCUSSION TITLE
Business Card Core Architecture

### DISCUSSION THREAD

**Discussion by TesteurManiak** [Floating-Dartists/fl_business_card_core#3] [Design Documentation]
This subproject of [fl_business_card ](https://github.com/Floating-Dartists/fl_business_card) and [fl_business_card_web](https://github.com/Floating-Dartists/fl_business_card_web) aims to unify some of the business logic common to both platforms. This code should be written in pure Dart and shouldn't have any dependency to Flutter-related libraries.

The base functionalities are (this might evolve) :

* User URL generation
* URL parsing and recognition

## User URL generation

We have to be able to generate custom URLs to generate the user's business card dynamically. Both the mobile app or the web app should be able to generate User URL to share.

## URL parsing and recognition

We have to be able to recognize some specific URLs to extract their relevant data. This will help us in the URL generation as we will be able to recompose an URL from the bare minimum data it should have.

### Proposition

[MeixDev's proposition will be the one we go with](https://github.com/Floating-Dartists/fl_business_card_core/discussions/3#discussioncomment-2253881)

## Supported URL pattern

For now we are aiming to support the following patterns in terms of recognition:

* Twitter
* LinkedIn
* GitHub
* GitLab
* StackOverflow
* Medium

### Parsing

When registering its URLs the user should be able to input either the full link to its profile or just its unique id.

### Technical guidance

Each URL pattern should be associated with a unique key to identify it (the less character the better):

* Twitter: `t`
* LinkedIn: `li`
* GitHub: `gh`
* GitLab: `gl`
* StackOverflow: `so`
* Medium: `m`
* Custom: `c`

The objective would be that because we can identify in advance the pattern we won't need to pass the entire thing when we are generating a "User URL".

### Example

If I want to generate a button to share my GitHub and StackOverflow profile it would generate the following query parameters:

```
https://my-super-cool-business-card/#/?gh=TesteurManiak&so=9942346
```

The part we are generating is the `gh=TesteurManiak&so=9942346`.

**Comment by MeixDev** [comment]
Current evolution follows the following principles :

## URL Parsing and recognition

### Proposition

We still use a base class `UrlParser`, which, as a pure abstract class would look like this:

```dart
abstract class UrlParser extends Equatable {
  const UrlParser({
    required this.service,
    this.schemes = const ["https", "http", ""],
    required this.hosts,
    this.pathSegments = const [],
    this.queryParameters = const {},
  });

  /// Name of the service the parser is for.
  ///
  /// Will be used in Exceptions, notably.
  final String service;

  /// The list of valid URL schemes for this service.
  ///
  /// As a rule of thumb, the first element of the list
  /// should be the most commonly used one, as it will
  /// be used to create the final URL with [recreateUri()]
  /// (aka `https` in most cases).
  ///
  /// For HTTP(S) schemes, it should also contain the empty string.
  final List<String> schemes;

  /// The list of valid URL hosts for this service.
  ///
  /// As a rule of thumb, the first element of the list
  /// should be the most commonly used one, as it will
  /// be used to create the final URL with [recreateUri()].
  ///
  /// In the case the username can be part of the host, the second
  /// host should be the apex host, so we can check again.
  final List<String> hosts;

  /// A list of the pathSegments that will be used in the URI.
  ///
  /// If a path segment should be replaced with the username,
  /// it should be called `{user}` in the list.
  ///
  /// The list shouldn't be used if it's empty.
  final List<String> pathSegments;

  /// A list of the queryParameters that will be used in the URI.
  ///
  /// If a query parameter should be replaced with the username,
  /// it should be called `{user}` in the map.
  ///
  /// The map shouldn't be used if it's empty.
  final Map<String, String> queryParameters;

  @override
  List<Object?> get props => [
        service,
        schemes,
        hosts,
        pathSegments,
        queryParameters,
      ];

  bool isValid(String uriString) => isValidUrl(uriString) == null;

  @mustCallSuper
  String? isValidUrl(String uriString);

  String recoverUser(String uriString);

  String recreateUri(String user);
}
```

However, all the functions are already implemented in this base function.

`isValidUrl()` checks for simple things, such as "is the URL part of the schemes and hosts". It is intended to be overriden by specific conditions in the extended classes, but they should always call their ancestor, hence the annotation.

The objective still is to keep as little as possible as we can from the URL, to get around the limitations of the URL & especially of the QR Code formats.
